language/vm/vm_runtime/src/substitution_map.rs

```diff
@@ -0,0 +1,102 @@
+use vm::{
+    errors::VMInvariantViolation,
+    file_format::{SignatureToken, StructHandleIndex, TypeParameterIndex},
+};
+
+#[cfg(test)]
+#[path = "unit_tests/substitution_map_tests.rs"]
+mod substitution_map_tests;
+
+enum SignatureTokenRef<'a> {
+    /// Boolean, `true` or `false`.
+    Bool,
+    /// Unsigned integers, 64 bits length.
+    U64,
+    /// Strings, immutable, utf8 representation.
+    String,
+    /// ByteArray, variable size, immutable byte array.
+    ByteArray,
+    /// Address, a 32 bytes immutable type.
+    Address,
+    /// MOVE user type, resource or unrestricted
+    Struct(StructHandleIndex, Vec<SignatureTokenRef<'a>>),
+    /// Type parameter.
+    TypeParameter(&'a SignatureTokenRef<'a>),
```

Comment:
This seems overly complicated, why not just use `SignatureToken`?

Reply:
I was trying to make copies only when necessary(when we try to materialize it)

Reply:
But why? I feel like you are going to need to copy this almost immediately.

Additionally, this adds an unusual layer of complexity, before we have any performance information

Reply:
I would agree with @tnowacki here that the effort seems too much.
I would start with using just SignatureToken and then see where it goes